server/controllers/finance/fiscal.js

```diff
@@ -457,3 +459,218 @@ function notNullBalance(array) {
     return (item.debit !== 0 || item.credit !== 0);
   });
 }
+
+/**
+ * @function closing
+ * @description closing a fiscal year
+ */
+function closing(req, res, next) {
+  let id = req.params.id,
+      accountId = req.body.params.account_id,
+      exploitation = {},
+      resultat  = {},
+      fiscal = {},
+      period = {},
+      sql = null;
+
+  // query fiscal year
+  sql = `SELECT id, number_of_months, end_date FROM fiscal_year WHERE id = ?;`;
+
+  db.exec(sql, [id])
+  .then(rows => {
+    if (!rows) {
+      throw new NotFound(`Could not find the fiscal year with id ${id}.`);
+    }
+
+    fiscal = rows[0];
+
+    // query period
+    sql = `SELECT p.id FROM period p WHERE p.fiscal_year_id = ? AND p.number = ?;`;
+    return db.exec(sql, [id, fiscal.number_of_months]);
+  })
+  .then(rows => {
+    if (!rows) {
+      throw new NotFound(`Could not find the period for the fiscal year with id ${id} and number ${fiscal.number_of_months}.`);
+    }
+    period = rows[0];
+  })
+  .then(() => {
+
+    let sqlProfitAccounts = `
+      SELECT a.id, pt.credit, pt.debit
+      FROM period_total AS pt
+      JOIN account AS a ON pt.account_id = a.id
+      JOIN account_type AS at ON a.type_id = at.id
+      WHERE (pt.fiscal_year_id = ? AND pt.period_id = ?) AND at.type = 'income';
+    `;
+    let sqlChargeAccounts = `
+      SELECT a.id, pt.credit, pt.debit
+      FROM period_total AS pt
+      JOIN account AS a ON pt.account_id = a.id
+      JOIN account_type AS at ON a.type_id = at.id
+      WHERE (pt.fiscal_year_id = ? AND pt.period_id = ?) AND at.type = 'income';
+    `;
+
+    let dbPromise = [
+      db.exec(sqlProfitAccounts, [id, period.id]),
+      db.exec(sqlChargeAccounts, [id, period.id])
+    ];
+    return q.all(dbPromise);
+  })
+  .spread((profitAccounts, chargeAccounts) => {
+
+    exploitation.profit = notNullBalance(profitAccounts);
+    exploitation.charge = notNullBalance(chargeAccounts);
+
+    // profit
+    resultat.profit = exploitation.profit.reduce((a, b) => {
+      return a + b.credit - b.debit;
+    }, 0);
+
+    // charge
+    resultat.charge = exploitation.charge.reduce((a, b) => {
+      return a + b.debit - b.credit;
+    }, 0);
+
+    // result
+    resultat.global = resultat.profit - resultat.charge;
+
+    let sqlTransId = `SELECT GenerateTransactionId(?) AS transId;`;
```

Comment:
Although it is always better to wrap this in a transaction to make sure it executes synchronously.  I propose that you start a transaction that does this and the `INSERT` calls.   What do you think?